File: Unit_tests_python/EX-03-zipf_law/Bard.py
from collections import Counter

def zipfs_law(text):
 """
 Finds the top 10 most frequent words in a text using Zipf's Law.

 Args:
     text: A string containing the text to analyze.

 Returns:
     A list of the top 10 most frequent words in the text, ordered by decreasing frequency.
 """

 # Lowercase the text and remove punctuation.
 text = text.lower()
 text = "".join(char for char in text if char.isalnum() or char.isspace())

 # Split the text into words.
 words = text.split()

 # Count the frequency of each word.
 word_counts = Counter(words)

 # Sort the word counts by frequency in descending order.
 most_frequent_words = word_counts.most_common(10)

 # Extract the words from the tuples.
 top_10_words = [word for word, _ in most_frequent_words]

 return top_10_words

File: Unit_tests_lua/EX-03-zipf_law/tests/Bard.lua
function zipfs_law(text)
    -- Remove punctuation and convert to lowercase
    local clean_text = string.gsub(text, "%p", "")
    clean_text = string.lower(clean_text)
  
    -- Split the text into words
    local words = string.split(clean_text)
  
    -- Create a table to store word counts
    local word_counts = {}
    for _, word in ipairs(words) do
      word_counts[word] = (word_counts[word] or 0) + 1
    end
  
    -- Sort the table by word count in descending order
    local sorted_words = {}
    for word, count in pairs(word_counts) do
      table.insert(sorted_words, {word = word, count = count})
    end
    table.sort(sorted_words, function(a, b) return b.count - a.count end)
  
    -- Return the top 10 most frequent words
    return table.sub(sorted_words, 1, 10)
  end  